grep.md

```diff
@@ -1,25 +1,60 @@
-After reading the entire file, return a list of all such pairs sorted
-from lowest line number to highest.
-
-The name "grep" comes from the [Unix program](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) with the same name.
-(In Unix, "grep" was a mnemonic for a useful command to another program [sed](http://www.gnu.org/software/sed/manual/sed.html), whose name means "stream editor".)
-
-The Unix `grep` can be run from the command line and accepts a variety of flags.
-Try reproducing some of these in your implementation. For example:
-- Accept a pattern and list of files on the command-line, print results
-  for each file to the console.
-- Implement the flags:
-  - `-l` Print only the names of files that contain at least one matching line.
-  - `-v` Invert the program -- collect all lines that fail to match the pattern.
-  - `-x` Only match entire lines, instead of lines that contain a match.
-- Support multiple flags at once.
-  For example, running `grep -l -x "hello" file1.txt file2.txt` should
-  print the names of files that do not contain the string "hello"
+The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command can be used to search for lines in one or more files 
+that match a user-provided search query.
+
+The `grep` command takes three arguments:
+
+1. The regular expression used to match lines in a file. 
+2. Zero or more flags to customize the matching behavior.
+3. One of more files in which to search for matching lines. 
+
+Your task is to implement the `grep` function, which should read the contents
+of the specified files, find the lines that match the specified pattern
+and then output those lines as a single string. Note that the lines should
+be output in the order in which they were found, with the first matching line
+in the first file being output first.
+
+As an example, suppose there is a file named "input.txt" with the following contents:
+
+<pre>
+hello
+world
+hello again
+</pre>
+
+If we were to call `grep "hello" input.txt`, the returned string should be:
+
+<pre>
+hello
+hello again
+</pre>
+
+### Flags
+
+As said earlier, the `grep` command should also support the following flags:
+
+- `-n` Print the line numbers of each matching line.
+- `-l` Print only the names of files that contain at least one matching line.
+- `-i` Match line using a case-insensitive comparison.
+- `-v` Invert the program -- collect all lines that fail to match the pattern.
+- `-x` Only match entire lines, instead of lines that contain a match.
+
+If we run `grep -n "hello" input.txt`, the `-n` flag will require the matching
+lines to be prefixed with its line number:
+
+<pre>
+1:hello
+3:hello again
+</pre>
+
+And if we run `grep -i "HELLO" input.txt`, we'll do a case-insensitive match, 
+and the output will be:
+
+<pre>
+hello
+hello again
+</pre>
+
+The `grep` command should support multiple flags at once.
+
+For example, running `grep -l -x "hello" file1.txt file2.txt` should
```

Comment:
either the `-x` should be a `-v`, or the `do not contain the string "hello"` should be changed to match what `-x` does